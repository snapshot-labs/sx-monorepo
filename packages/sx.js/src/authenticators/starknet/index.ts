import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import createVanillaAuthenticator from './vanilla';
import { Authenticator } from '../../clients/starknet/types';
import { NetworkConfig } from '../../types';
import { hexPadLeft } from '../../utils/encoding';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
