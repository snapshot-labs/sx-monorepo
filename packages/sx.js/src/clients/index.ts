export { EthereumTx } from './starknet/ethereum-tx';
export { EthereumSig } from './starknet/ethereum-sig';
export { StarknetTx } from './starknet/starknet-tx';
export { StarknetSig } from './starknet/starknet-sig';
export { L1Executor } from './starknet/l1-executor';
export { HerodotusController } from './starknet/herodotus';
export { EthereumTx as EvmEthereumTx } from './evm/ethereum-tx';
export { EthereumSig as EvmEthereumSig } from './evm/ethereum-sig';
export { EthereumTx as GovernorBravoEthereumTx } from './governor-bravo/ethereum-tx';
export { EthereumSig as OffchainEthereumSig } from './offchain/ethereum-sig';
export { StarknetSig as OffchainStarknetSig } from './offchain/starknet-sig';
