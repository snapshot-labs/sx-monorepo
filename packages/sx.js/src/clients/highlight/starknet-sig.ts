import { Account, StarknetDomain, StarknetType, TypedData } from 'starknet';
import { Envelope, HighlightDomain, SetAlias } from './types';
import {
  HIGHLIGHT_STARKNET_DOMAIN,
  STARKNET_ALIASES_CONFIG
} from '../../highlightConstants';

export class HighlightStarknetSigClient {
  private highlightUrl: string;

  constructor(highlightUrl: string) {
    this.highlightUrl = highlightUrl;
  }

  public async getDomain(
    signer: Account,
    salt: bigint,
    to: string
  ): Promise<{
    starknetDomain: Required<StarknetDomain>;
    meta: Pick<HighlightDomain, 'salt' | 'verifyingContract'>;
  }> {
    return {
      starknetDomain: {
        ...HIGHLIGHT_STARKNET_DOMAIN,
        chainId: await signer.getChainId()
      },
      meta: {
        salt: `0x${salt.toString(16).padStart(64, '0')}`,
        verifyingContract: to
      }
    };
  }

  public async sign(
    signer: Account,
    domain: Required<StarknetDomain>,
    primaryType: string,
    types: Record<string, StarknetType[]>,
    message: Record<string, any>
  ): Promise<string> {
    const data: TypedData = {
      types,
      primaryType,
      domain,
      message
    };

    const signature = await signer.signMessage(data);

    return (
      Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`]
    ).join(',');
  }

  public async send(envelope: Envelope): Promise<any> {
    const { domain, message, primaryType, signer, signature } = envelope;

    const payload = {
      method: 'hl_postMessage',
      params: {
        domain,
        message,
        primaryType,
        signer,
        signature
      }
    };

    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    };

    const res = await fetch(this.highlightUrl, body);

    if (!res.ok) {
      throw new Error(`Failed to send envelope: ${res.statusText}`);
    }

    const result = await res.json();

    if (result.error) {
      throw new Error(result.error);
    }

    return result;
  }

  public async setAlias({
    signer,
    data,
    salt
  }: {
    signer: Account;
    data: SetAlias;
    salt: bigint;
  }): Promise<Envelope> {
    const primaryType = 'SetAlias';

    const domain = await this.getDomain(
      signer,
      salt,
      STARKNET_ALIASES_CONFIG.address
    );

    const signature = await this.sign(
      signer,
      domain.starknetDomain,
      primaryType,
      STARKNET_ALIASES_CONFIG.types.setAlias,
      data
    );

    return {
      type: 'HIGHLIGHT_ENVELOPE',
      domain: { ...domain.starknetDomain, ...domain.meta },
      message: data,
      primaryType,
      signer: signer.address,
      signature
    };
  }
}
