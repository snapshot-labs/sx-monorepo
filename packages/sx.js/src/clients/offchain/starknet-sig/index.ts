import { Account, TypedData, validateAndParseAddress } from 'starknet';
import { aliasTypes, baseDomain } from './types';
import { OffchainNetworkStarknetConfig } from '../../../types';
import {
  EIP712SetAliasMessage,
  Envelope,
  SetAlias,
  SignatureData
} from '../types';

type StarknetSigClientOpts = {
  networkConfig: OffchainNetworkStarknetConfig;
};

export class StarknetSig {
  config: StarknetSigClientOpts;

  constructor(opts: StarknetSigClientOpts) {
    this.config = opts;
  }

  public async sign<T extends EIP712SetAliasMessage>(
    signer: Account,
    message: T,
    types: any,
    primaryType: string
  ): Promise<SignatureData> {
    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId
    };

    const data: TypedData = {
      types,
      primaryType,
      domain,
      message
    };
    const signature = await signer.signMessage(data);

    return {
      address: validateAndParseAddress(signer.address),
      signature: Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`],
      message,
      domain,
      types,
      primaryType
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Account;
    data: SetAlias;
  }): Promise<Envelope<SetAlias>> {
    const message = {
      ...data,
      from: validateAndParseAddress(signer.address),
      timestamp: parseInt((Date.now() / 1000).toFixed())
    };

    const signatureData = await this.sign(
      signer,
      message,
      aliasTypes,
      'SetAlias'
    );

    return {
      signatureData,
      data
    };
  }
}
