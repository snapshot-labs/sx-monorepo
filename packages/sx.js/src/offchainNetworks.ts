import { constants as starknetConstants } from 'starknet';
import {
  OffchainNetworkEthereumConfig,
  OffchainNetworkStarknetConfig
} from './types';

type EvmChainId = OffchainNetworkEthereumConfig['eip712ChainId'];
type StarknetChainId = OffchainNetworkStarknetConfig['eip712ChainId'];

function createStandardConfig<T>(eip712ChainId: T) {
  return {
    eip712ChainId
  };
}

export const offchainMainnet = createStandardConfig<EvmChainId>(1);
export const offchainGoerli = createStandardConfig<EvmChainId>(5);
export const offchainStarknetMainnet = createStandardConfig<StarknetChainId>(
  starknetConstants.StarknetChainId.SN_MAIN
);
export const offchainStarknetSepolia = createStandardConfig<StarknetChainId>(
  starknetConstants.StarknetChainId.SN_SEPOLIA
);
