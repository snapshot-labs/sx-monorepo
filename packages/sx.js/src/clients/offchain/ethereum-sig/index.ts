import {
  Signer,
  TypedDataField,
  TypedDataSigner
} from '@ethersproject/abstract-signer';
import {
  aliasTypes,
  approvalVoteTypes,
  basicVoteTypes,
  cancelProposalTypes,
  deleteSpaceTypes,
  domain,
  encryptedVoteTypes,
  followSpaceTypes,
  proposeTypes,
  rankedChoiceVoteTypes,
  singleChoiceVoteTypes,
  unfollowSpaceTypes,
  updateProposalTypes,
  updateSpaceTypes,
  updateStatementTypes,
  updateUserTypes,
  weightedVoteTypes
} from './types';
import { offchainGoerli } from '../../../offchainNetworks';
import { OffchainNetworkConfig, SignatureData } from '../../../types';
import {
  CancelProposal,
  DeleteSpace,
  EIP712CancelProposalMessage,
  EIP712DeleteSpaceMessage,
  EIP712FollowSpaceMessage,
  EIP712Message,
  EIP712ProposeMessage,
  EIP712SetAliasMessage,
  EIP712UnfollowSpaceMessage,
  EIP712UpdateProposal,
  EIP712UpdateSpaceMessage,
  EIP712UpdateStatementMessage,
  EIP712UpdateUserMessage,
  EIP712VoteMessage,
  Envelope,
  FollowSpace,
  Propose,
  SetAlias,
  UnfollowSpace,
  UpdateProposal,
  UpdateSpace,
  UpdateStatement,
  UpdateUser,
  Vote
} from '../types';
import { encryptChoices } from '../utils';

const SEQUENCER_URLS: Record<OffchainNetworkConfig['eip712ChainId'], string> = {
  1: 'https://seq.snapshot.org',
  5: 'https://testnet.seq.snapshot.org'
};

const RELAYER_URLS: Record<OffchainNetworkConfig['eip712ChainId'], string> = {
  1: 'https://relayer.snapshot.org',
  5: 'https://testnet.seq.snapshot.org' // no relayer for testnet
};

type EthereumSigClientOpts = {
  networkConfig?: OffchainNetworkConfig;
  sequencerUrl?: string;
};

export class EthereumSig {
  sequencerUrl: string;
  networkConfig: OffchainNetworkConfig;

  constructor(opts?: EthereumSigClientOpts) {
    this.networkConfig = opts?.networkConfig || offchainGoerli;
    this.sequencerUrl =
      opts?.sequencerUrl || SEQUENCER_URLS[this.networkConfig.eip712ChainId];
  }

  public async sign<
    T extends
      | EIP712VoteMessage
      | EIP712ProposeMessage
      | EIP712UpdateProposal
      | EIP712CancelProposalMessage
      | EIP712FollowSpaceMessage
      | EIP712UnfollowSpaceMessage
      | EIP712SetAliasMessage
      | EIP712UpdateUserMessage
      | EIP712UpdateStatementMessage
      | EIP712UpdateSpaceMessage
      | EIP712DeleteSpaceMessage
  >(
    signer: Signer & TypedDataSigner,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();
    const EIP712Message: EIP712Message = {
      from: address,
      timestamp: parseInt((Date.now() / 1000).toFixed()),
      ...message
    };
    const signature = await signer._signTypedData(domain, types, EIP712Message);
    return {
      address,
      signature,
      domain,
      types,
      message: EIP712Message
    };
  }

  public async send(
    envelope: Envelope<
      | Vote
      | Propose
      | UpdateProposal
      | CancelProposal
      | FollowSpace
      | UnfollowSpace
      | SetAlias
    >
  ) {
    const {
      address,
      signature: sig,
      domain,
      types,
      primaryType,
      message
    } = envelope.signatureData!;
    const payload: any = {
      address,
      sig,
      data: {
        domain,
        types,
        message
      }
    };

    // primaryType needs to be attached when sending starknet-sig generated payload
    if (primaryType) payload.data.primaryType = primaryType;

    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    };

    let url = this.sequencerUrl;
    if (sig === '0x') {
      url = RELAYER_URLS[this.networkConfig.eip712ChainId] || url;
    }
    const res = await fetch(url, body);
    const result = await res.json();

    if (result.error) {
      throw new Error(
        typeof result.error_description === 'string'
          ? result.error_description
          : result.error
      );
    }

    return result;
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const signatureData = await this.sign(signer, data, proposeTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const signatureData = await this.sign(signer, data, updateProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async cancel({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: CancelProposal;
  }): Promise<Envelope<CancelProposal>> {
    const signatureData = await this.sign(signer, data, cancelProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    let choice: EIP712VoteMessage['choice'];
    let voteType: { Vote: { name: string; type: string }[] };

    switch (data.type) {
      case 'single-choice':
        voteType = singleChoiceVoteTypes;
        choice = data.choice as number;
        break;
      case 'approval':
        voteType = approvalVoteTypes;
        choice = data.choice as number[];
        break;
      case 'ranked-choice':
        voteType = rankedChoiceVoteTypes;
        choice = data.choice as number[];
        break;
      case 'weighted':
      case 'quadratic':
        voteType = weightedVoteTypes;
        choice = JSON.stringify(data.choice);
        break;
      default:
        voteType = basicVoteTypes;
        choice = data.choice as number;
    }

    const message: EIP712VoteMessage = {
      space: data.space,
      proposal: data.proposal,
      choice,
      reason: data.reason || '',
      app: data.app,
      metadata: ''
    };

    if (data.privacy) {
      message.privacy = data.privacy;
      voteType = encryptedVoteTypes;
      message.choice = await encryptChoices(
        data.privacy,
        data.proposal,
        typeof message.choice === 'string'
          ? message.choice
          : JSON.stringify(message.choice)
      );
    }
    const signatureData = await this.sign(signer, message, voteType);

    return {
      signatureData,
      data
    };
  }

  public async followSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: FollowSpace;
  }): Promise<Envelope<FollowSpace>> {
    const signatureData = await this.sign(signer, data, followSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async unfollowSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UnfollowSpace;
  }): Promise<Envelope<UnfollowSpace>> {
    const signatureData = await this.sign(signer, data, unfollowSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: SetAlias;
  }): Promise<Envelope<SetAlias>> {
    const signatureData = await this.sign(signer, data, aliasTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateUser({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateUser;
  }) {
    const signatureData = await this.sign(signer, data, updateUserTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateStatement({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateStatement;
  }) {
    const signatureData = await this.sign(signer, data, updateStatementTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateSpace;
  }) {
    const signatureData = await this.sign(signer, data, updateSpaceTypes);

    return {
      signatureData,
      data
    };
  }

  public async deleteSpace({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: DeleteSpace;
  }) {
    const signatureData = await this.sign(signer, data, deleteSpaceTypes);

    return {
      signatureData,
      data
    };
  }
}
