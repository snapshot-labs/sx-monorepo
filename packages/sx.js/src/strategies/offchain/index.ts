import createOnlyMembersStrategy from './only-members';
import createRemoteValidateStrategy from './remote-validate';
import createRemoteVpStrategy from './remote-vp';
import { Strategy } from '../../clients/offchain/types';

export function getStrategy(name: string): Strategy | null {
  if (name === 'only-members') return createOnlyMembersStrategy();

  if (['any', 'basic', 'passport-gated', 'arbitrum', 'karma-eas-attestation'].includes(name)) {
    return createRemoteValidateStrategy(name);
  }

  return createRemoteVpStrategy();
}
