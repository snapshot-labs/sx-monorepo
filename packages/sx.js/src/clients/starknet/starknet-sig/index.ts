import randomBytes from 'randombytes';
import {
  Account,
  CallData,
  shortString,
  typedData,
  uint256,
  validateAndParseAddress
} from 'starknet';
import {
  aliasTypes,
  baseDomain,
  proposeTypes,
  updateProposalTypes,
  voteTypes
} from './types';
import {
  Alias,
  ClientConfig,
  ClientOpts,
  Envelope,
  Propose,
  SignatureData,
  StarknetEIP712AliasMessage,
  StarknetEIP712ProposeMessage,
  StarknetEIP712UpdateProposalMessage,
  StarknetEIP712VoteMessage,
  UpdateProposal,
  Vote
} from '../../../types';
import { getStrategiesWithParams } from '../../../utils/strategies';

export class StarknetSig {
  config: ClientConfig & { manaUrl: string };

  constructor(opts: ClientOpts & { manaUrl: string }) {
    this.config = opts;
  }

  generateSalt() {
    return `0x${randomBytes(4).toString('hex')}`;
  }

  public async send(envelope: Envelope<Propose | UpdateProposal | Vote>) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(
      `${this.config.manaUrl}/stark_rpc/${this.config.networkConfig.eip712ChainId}`,
      body
    );
    const json = await res.json();

    return json.result;
  }

  public async sign<
    T extends
      | StarknetEIP712ProposeMessage
      | StarknetEIP712UpdateProposalMessage
      | StarknetEIP712VoteMessage
      | StarknetEIP712AliasMessage
  >(
    signer: Account,
    verifyingContract: string,
    message: T,
    types: any,
    primaryType: string
  ): Promise<SignatureData> {
    const domain = {
      ...baseDomain,
      chainId: this.config.networkConfig.eip712ChainId,
      verifyingContract
    };

    const data: typeof typedData.TypedData = {
      types,
      primaryType,
      domain,
      message
    };
    const signature = await signer.signMessage(data);

    return {
      address: validateAndParseAddress(signer.address),
      signature: Array.isArray(signature)
        ? signature.map(v => `0x${BigInt(v).toString(16)}`)
        : [`0x${signature.r.toString(16)}`, `0x${signature.s.toString(16)}`],
      message,
      domain,
      types,
      primaryType
    };
  }

  public async propose({
    signer,
    data
  }: {
    signer: Account;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const address = signer.address;

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      author: address,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      userProposalValidationParams: CallData.compile({
        userStrategies
      }),
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      proposeTypes,
      'Propose'
    );

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Account;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const address = signer.address;

    const message = {
      space: data.space,
      author: address,
      proposalId: uint256.bnToUint256(data.proposal),
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      updateProposalTypes,
      'UpdateProposal'
    );

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Account;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const address = signer.address;

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      space: data.space,
      voter: address,
      proposalId: uint256.bnToUint256(data.proposal),
      choice: `0x${data.choice.toString(16)}`,
      userVotingStrategies,
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str))
    };

    const signatureData = await this.sign(
      signer,
      data.authenticator,
      message,
      voteTypes,
      'Vote'
    );

    return {
      signatureData,
      data
    };
  }

  public async setAlias({
    signer,
    data
  }: {
    signer: Account;
    data: Alias;
  }): Promise<Envelope<Alias>> {
    const message = {
      from: validateAndParseAddress(signer.address),
      timestamp: parseInt((Date.now() / 1000).toFixed()),
      ...data
    };

    const signatureData = await this.sign(
      signer,
      '',
      message,
      aliasTypes,
      'SetAlias'
    );

    return {
      signatureData,
      data
    };
  }
}
