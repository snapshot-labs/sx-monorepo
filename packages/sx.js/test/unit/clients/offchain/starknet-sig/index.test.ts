import { Account } from 'starknet';
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { StarknetSig } from '../../../../../src/clients/offchain/starknet-sig';
import { offchainStarknetSepolia } from '../../../../../src/offchainNetworks';
import { starkProvider } from '../../../helpers';

describe('StarknetSig', () => {
  const address =
    '0x7d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4';
  const privateKey = '0xcd613e30d8f16adf91b7584a2265b1f5';

  const account = new Account(starkProvider, address, privateKey);

  const client = new StarknetSig({ networkConfig: offchainStarknetSepolia });

  beforeAll(() => {
    vi.useFakeTimers({
      now: new Date('2024-01-21').getTime()
    });
  });

  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('should create setAlias envelope', async () => {
    const envelope = await client.setAlias({
      signer: account,
      data: {
        alias: '0xdeadbeef'
      }
    });

    expect(envelope).toMatchSnapshot();
  });
});
