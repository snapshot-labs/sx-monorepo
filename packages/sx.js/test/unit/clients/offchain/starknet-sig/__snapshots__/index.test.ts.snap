// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StarknetSig > should create setAlias envelope 1`] = `
{
  "data": {
    "alias": "0xdeadbeef",
  },
  "signatureData": {
    "address": "0x07d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
    "domain": {
      "chainId": "0x534e5f5345504f4c4941",
      "name": "sx-starknet",
      "version": "0.1.0",
    },
    "message": {
      "alias": "0xdeadbeef",
      "from": "0x07d2f37b75a5e779f7da01c22acee1b66c39e8ba470ee5448f05e1462afcedb4",
      "timestamp": 1705795200,
    },
    "primaryType": "SetAlias",
    "signature": [
      "0x41c602fccc95b2a605a0a5d5ccbd42eabb9d5d7a09e436b3cddd1fdc1b5310d",
      "0x64bed5315abdc4e2e78bd84c9cfa13cd17a5e9c370c4a89514b754b693ea9d5",
    ],
    "types": {
      "SetAlias": [
        {
          "name": "from",
          "type": "felt",
        },
        {
          "name": "alias",
          "type": "felt",
        },
        {
          "name": "timestamp",
          "type": "felt",
        },
      ],
      "StarkNetDomain": [
        {
          "name": "name",
          "type": "felt",
        },
        {
          "name": "version",
          "type": "felt",
        },
        {
          "name": "chainId",
          "type": "felt",
        },
      ],
    },
  },
}
`;
